### PR TITLE
Switch existing implementation of programming question form to React

### PR DESCRIPTION
**Pull request by justwudi**
Next step of #1568
Depends on #1798 

This PR:

- Implements the existing programming question form using ReactJS and material-ui.
- Adds `ChipInput`, a modified version of the material multi-select from [material-ui-chip-input](https://github.com/TeamWertarbyte/material-ui-chip-input).
- Change editing of existing programming question to asynchronous form submission
  + The import alert and build log will be hidden when loading and shown after evaluation.

After this:

- Implement the online editor for python.

### CODE REVIEW

File: app/views/course/assessment/question/programming/_question.json.jbuilder [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  json.staff_only_comments @programming_question.staff_only_comments
+  json.maximum_grade @programming_question.maximum_grade
+  json.weight @programming_question.weight
+  json.language_id @programming_question.language_id
+  json.languages Coursemology::Polyglot::Language.all do |lang|
+    json.id lang.id
+    json.name lang.name
+    json.editor_mode editor_mode(lang)
+  end
+  json.skill_ids @programming_question.skills.as_json(only: [:id, :title])
```
Comment (allenwq) [inline app/views/course/assessment/question/programming/_question.json.jbuilder]: The tags need to be sorted

File: app/views/course/assessment/question/programming/_form.html.slim [hunk 2]
```diff
@@ -1,24 +1,2 @@
...
-
-  = render partial: 'form_templates',
-           locals: { programming_question: @programming_question }
-
-  = render partial: 'form_test_cases',
-           locals: { test_cases: @programming_question.test_cases }
-
-  = f.button :submit
+div#programming-question
+  = fa_icon 'spinner 4x spin'
```
Comment (allenwq) [inline app/views/course/assessment/question/programming/_form.html.slim]: Don't use an extra request to load the form. You can still render the JSON props and make it the attribute of a div, and read the JSON from div in the frontend. 

Or you could find better ways to do that.

**Comment by allenwq** [comment]
@justwudi Aso the styling need to be improved. After material style comes, It's a bit inconsistent now (The test cases are still bootstrap )   HTML editor also need to be styled.

**Comment by allenwq** [comment]
@kxmbrian Do you have comments about the React code?

**Comment by kxmbrian** [comment]
The skills dropdown doesn't scroll. Can't see some of the items if the list is long.

File: app/views/course/assessment/question/programming/_form.html.slim (hunk 2 above)
Comment (kxmbrian) [inline app/views/course/assessment/question/programming/_form.html.slim]: Can take reference to #1800 as an example

**Comment by allenwq** [comment]
@justwudi Looks much better now ! Thanks !

Few more comments about the UI after your latest change:

- "Download package: " should be "Uploaded Package:", also need some spacing between the upload button and the "Download package" text, now they are too close.

- Should leave some space between the number inputs ( now there are kissing each other)

- Can fix the attempt limit hint overflow? I'd imagine memory limit and time limit also need some hints later.

File: app/views/course/assessment/question/programming/_question.json.jbuilder (hunk 1 above)
Comment (allenwq) [inline app/views/course/assessment/question/programming/_question.json.jbuilder]: @justwudi The tags are still not sorted.

**Comment by allenwq** [comment]
👍

**Pull request merged**